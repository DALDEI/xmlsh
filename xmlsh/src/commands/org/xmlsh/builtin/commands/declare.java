/**
 * $Id: colon.java 21 2008-07-04 08:33:47Z daldei $
 * $Date: 2008-07-04 04:33:47 -0400 (Fri, 04 Jul 2008) $
 *
 */

package org.xmlsh.builtin.commands;

import static org.xmlsh.core.XVariable.XVarFlag.EXPORT;
import static org.xmlsh.core.XVariable.XVarFlag.LOCAL;
import static org.xmlsh.core.XVariable.XVarFlag.READONLY;

import java.io.IOException;
import java.io.PrintWriter;
import java.io.UnsupportedEncodingException;
import java.util.EnumSet;
import java.util.List;
import java.util.Map.Entry;

import javax.xml.stream.XMLStreamException;
import javax.xml.stream.XMLStreamWriter;

import net.sf.saxon.s9api.SaxonApiException;

import org.xml.sax.SAXException;
import org.xmlsh.core.BuiltinCommand;
import org.xmlsh.core.CoreException;
import org.xmlsh.core.InvalidArgumentException;
import org.xmlsh.core.Options;
import org.xmlsh.core.XEnvironment;
import org.xmlsh.core.XValue;
import org.xmlsh.core.XVariable;
import org.xmlsh.core.XVariable.XVarFlag;
import org.xmlsh.core.io.OutputPort;
import org.xmlsh.sh.module.IModule;
import org.xmlsh.sh.shell.FunctionDefinitions;
import org.xmlsh.sh.shell.IFunctionDefiniton;
import org.xmlsh.sh.shell.SerializeOpts;
import org.xmlsh.util.Util;

public class declare extends BuiltinCommand {

	@Override
	public int run(List<XValue> args) throws Exception {
		Options opts = new Options("p,r,t,x,f,l,m,v,",
				SerializeOpts.getOptionDefs());
		opts.parse(args);
		setSerializeOpts(opts);
		args = opts.getRemainingArgs();
		boolean bFound = false ;
		boolean hasP = opts.hasOpt("p");
		boolean hasAny = 
				 opts.hasOpt("f") ||
				 opts.hasOpt("v") ||
				 opts.hasOpt("m") ||
				 opts.hasOpt("n") ;
		
		boolean hasType = 
				opts.hasOpt("x") ||
				opts.hasOpt("l") ||
				opts.hasOpt("f") ||
				opts.hasOpt("r") ;
		

		boolean hasArgs = ! args.isEmpty();
		boolean hasNS  = hasArgs && args.get(0).toString().equals("namespace");

		if( ! hasP && ! hasArgs && ! hasNS  )
			hasP = true ;
            
			
		if( ! hasType && ! hasNS )
			hasP = true ;

		if( hasP && (!hasAny || opts.hasOpt("f")))
			bFound |= printFuncs(opts, args);
		if (hasP &&  (!hasAny || opts.hasOpt("v"))) 
			bFound |=  printVars(opts, args );
		if (hasP &&  (!hasAny || opts.hasOpt("m"))) 
			bFound |=  printModules(opts, args);
		if (hasP &&  (!hasAny || opts.hasOpt("n")))
			bFound |=  printNamespaces(opts, args);
		
		
		
		
		if( ! hasP && hasArgs ) {

		
			if( hasType )
				return declareType( opts , args );
		    else
		    if( hasNS ){
				XValue what = args.remove(0);
				if (what.toString().equals("namespace")) {
					if( args.isEmpty() )
						printNamespaces(opts, args);
					else
						declareNamespace(args.get(0));
				}
				else 
					usage();
				return -1;
			}
		} 
		else 
		if( ! bFound && !  args.isEmpty() ){
		    mShell.printErr( args.get(0) + ": not found");
			return 1;
		}
			
		return 0;
		


	}

	private int declareType(Options opts, List<XValue> args) throws InvalidArgumentException {

		
		if( opts.hasOpt("f") ){
			mShell.printErr("predeclaring functions is not implemented.");
			return 1 ;
		}
		
		EnumSet<XVarFlag> flags = XVariable.standardFlags();
		
		if( opts.hasOpt("x") )
			flags = Util.withEnumAdded(flags, EXPORT);
		if( opts.hasOpt("l") )
			flags = Util.withEnumAdded(flags, LOCAL);
		if( opts.hasOpt("r") )
			flags = Util.withEnumAdded(flags, READONLY);
		
		for( XValue a : args )
			mShell.getEnv().declareVar(a.toString(), flags);
		return 0 ;
		
		
	}

	private boolean  printNamespaces(Options opts, List<XValue> args) {
		boolean bFound = false ;
		if (args.isEmpty())
			bFound = listNamespaces();
		else {
			for (XValue a : args)
				bFound |= listNamespace(a.toString());
		}

		return bFound ;

	}

	private boolean printVars(Options opts, List<XValue> args)
			throws UnsupportedEncodingException, CoreException, IOException,
			XMLStreamException, SAXException, SaxonApiException {
		if (args.size() < 1) {
			return printAllVars();

		} else
			return printVar(args.get(0).toString());
	}

	private boolean printFuncs(Options opts, List<XValue> args)
			throws UnsupportedEncodingException, CoreException, IOException {
		if (args.isEmpty())
			return printAllFuncs();
		else
			return printFunc(args.get(0).toString());
		
	}

	private boolean printAllVars() throws UnsupportedEncodingException,
			CoreException, IOException {

		try (PrintWriter w = getStdout().asPrintWriter(getSerializeOpts())) {

			for (String var : getShell().getEnv().getVarNames())
				w.println(var);
		}
		return true ;

	}

	private boolean  printAllFuncs() throws UnsupportedEncodingException,
			CoreException, IOException {
		FunctionDefinitions fd = getShell().getFunctionDelcs();
		if (fd == null)
			return true ;
		try (PrintWriter w = getStdout().asPrintWriter(getSerializeOpts())) {

			for (Entry<String, IFunctionDefiniton> e : getShell().getEnv()
					.getFunctions().entrySet()) {

				w.printf("%s [%s]\n", e.getKey(), e.getValue().getModule()
						.getName());

			}
		}
		return true  ;

	}

	private boolean printFunc(String name) throws UnsupportedEncodingException,
			CoreException, IOException {
		boolean bFound = false;
		try (PrintWriter w = getStdout().asPrintWriter(getSerializeOpts())) {

			IFunctionDefiniton fd = getShell().getFunctionDecl(name);
			if (fd != null) {
				bFound = true ;

				w.print(name);
				w.println("()");
				w.println(fd.getBody().describe(false));

			}
		}
		return bFound ;

	}


	private boolean printVar(String name) throws XMLStreamException, SAXException,
			IOException, CoreException, SaxonApiException {
		XEnvironment env = mShell.getEnv();

		XVariable var = env.getVar(name);
		if (var == null) {
			return false ; 
		}

		OutputPort stdout = env.getStdout();
		XMLStreamWriter writer = stdout.asXMLStreamWriter(getSerializeOpts());

		try {
			writer.writeStartDocument();
			writer.writeStartElement(getName());
			var.serialize(writer);
			writer.writeEndElement();
			writer.writeEndDocument();

		} finally {
			if (writer != null)
				writer.close();
			stdout.writeSequenceTerminator(getSerializeOpts());
		}

		return true ;

	}

	private int declareNamespace(List<XValue> args) {
		if (args.size() == 0)
			 listNamespaces();

		for (XValue arg : args) {
			declareNamespace(arg);

		}
		return 0;

	}

	private boolean listNamespace(String name) {
		String uri =  mShell.getEnv().getNamespaces().get(name);
		if (uri == null) 
			return false ;

		mShell.printOut(name + "=" + uri);

		return true ;
	}

	private boolean listNamespaces() {
		for (Entry<String, String> e :  mShell.getEnv().getNamespaces().entrySet() ) {
			mShell.printOut(e.getKey() + "=" + e.getValue() );
		}
		return true ;

	}

	/*
	 * Declare a namespace
	 */

	private int declareNamespace(XValue arg) {

		// ns="url"
		// ns=
		// "url"
		if (arg.isAtomic()) {

			mShell.getEnv().declareNamespace(arg.toString());

			return 0 ;

		} else
			usage("Unexpected value: " + arg.describe());
		return 2;
	}

	private boolean printModules(Options opts, List<XValue> args) {

		boolean bFound = false ;
		if (args.isEmpty()) {
			for (IModule hm : mShell.getEnv().getModules()) {
				printModule(hm);

			}
			return true ;
		} else {
			for (XValue a : args) {
				String name = a.toString();
				IModule hm = getEnv().getModules()
						.getExistingModuleByName(name);
				if( hm != null )
					bFound |= printModule(hm);
			}
		}
		return bFound ;

	}

	protected boolean  printModule(IModule hm) {
		assert( hm != null );
		
		XEnvironment env = getEnv();
		StringBuilder sb = new StringBuilder();
		String p = "";
		for (String prefix : env.getPrefixesForModule(hm)) {
			sb.append(p).append(prefix).append('=');
			p = " ";
		}

		String s = hm.describe();
		if (Util.isBlank(s))
			sb.append(hm.getName());
		else
			sb.append(s);

		mShell.printOut(sb.toString());
		return true ;
	}

}
//
//
// Copyright (C) 2008-2014 David A. Lee.
//
// The contents of this file are subject to the "Simplified BSD License" (the
// "License");
// you may not use this file except in compliance with the License. You may
// obtain a copy of the
// License at http://www.opensource.org/licenses/bsd-license.php
//
// Software distributed under the License is distributed on an "AS IS" basis,
// WITHOUT WARRANTY OF ANY KIND, either express or implied.
// See the License for the specific language governing rights and limitations
// under the License.
//
// The Original Code is: all this file.
//
// The Initial Developer of the Original Code is David A. Lee
//
// Portions created by (your name) are Copyright (C) (your legal entity). All
// Rights Reserved.
//
// Contributor(s): none.
//
