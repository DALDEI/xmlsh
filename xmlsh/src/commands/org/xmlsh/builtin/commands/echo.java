/**
 * $Id: echo.java 88 2008-11-27 17:06:00Z daldei $
 * $Date: 2008-11-27 12:06:00 -0500 (Thu, 27 Nov 2008) $
 *
 */

package org.xmlsh.builtin.commands;

import java.io.OutputStream;
import java.util.List;
import org.xmlsh.core.BuiltinCommand;
import org.xmlsh.core.InvalidArgumentException;
import org.xmlsh.core.Options;
import org.xmlsh.core.XValue;
import org.xmlsh.core.io.OutputPort;
import org.xmlsh.sh.shell.SerializeOpts;
import org.xmlsh.util.Util;

public class echo extends BuiltinCommand {

  @Override
  public int run(List<XValue> args) throws Exception {

    Options opts = new Options("n,p=port:", SerializeOpts.getOptionDefs());
    opts.parse(args);

    boolean nolf = opts.hasOpt("n");
    String port = opts.getOptString("p", null);

    OutputPort stdout = port != null ? mShell.getEnv().getOutputPort(port)
        : mShell.getEnv().getStdout();

    if(stdout == null)
      throw new InvalidArgumentException("Output port not found: " + port);

    SerializeOpts serializeOpts = getSerializeOpts(opts);

    OutputStream out = stdout.asOutputStream(serializeOpts);

    args = opts.getRemainingArgs();

    // args = Util.expandSequences( args);
    boolean bFirst = true;
    for(XValue arg : args) {
      if(!bFirst)
        out.write(' ');

      bFirst = false;
      arg.serialize(out, serializeOpts);
    }
    if(!nolf)
      out.write(Util.getNewlineBytes(serializeOpts));
    out.flush();
    return 0;
  }
}
//
//
// Copyright (C) 2008-2014 David A. Lee.
//
// The contents of this file are subject to the "Simplified BSD License" (the
// "License");
// you may not use this file except in compliance with the License. You may
// obtain a copy of the
// License at http://www.opensource.org/licenses/bsd-license.php
//
// Software distributed under the License is distributed on an "AS IS" basis,
// WITHOUT WARRANTY OF ANY KIND, either express or implied.
// See the License for the specific language governing rights and limitations
// under the License.
//
// The Original Code is: all this file.
//
// The Initial Developer of the Original Code is David A. Lee
//
// Portions created by (your name) are Copyright (C) (your legal entity). All
// Rights Reserved.
//
// Contributor(s): none.
//
