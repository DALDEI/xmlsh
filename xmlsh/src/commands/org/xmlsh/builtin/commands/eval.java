/**
 * $Id: colon.java 21 2008-07-04 08:33:47Z daldei $
 * $Date: 2008-07-04 04:33:47 -0400 (Fri, 04 Jul 2008) $
 *
 */

package org.xmlsh.builtin.commands;

import org.xmlsh.core.BuiltinCommand;
import org.xmlsh.core.XValue;
import org.xmlsh.sh.core.Command;
import org.xmlsh.sh.shell.ShellConstants;
import org.xmlsh.util.Util;

import java.util.List;

public class eval extends BuiltinCommand {

	/*
	 * Evaluate all arguments as as string
	 * and parse them as a command 
	 * 
	 */

	@Override
	public int run(  List<XValue> args ) throws Exception {
		String  evals = Util.joinValues(args,ShellConstants.ARG_SEPARATOR );
		Command c = mShell.parseEval(evals);
		if( c == null )
			return 0;

		int ret = mShell.exec(c);
		return ret ;
	}



}
//
//
//Copyright (C) 2008-2014    David A. Lee.
//
//The contents of this file are subject to the "Simplified BSD License" (the "License");
//you may not use this file except in compliance with the License. You may obtain a copy of the
//License at http://www.opensource.org/licenses/bsd-license.php 
//
//Software distributed under the License is distributed on an "AS IS" basis,
//WITHOUT WARRANTY OF ANY KIND, either express or implied.
//See the License for the specific language governing rights and limitations under the License.
//
//The Original Code is: all this file.
//
//The Initial Developer of the Original Code is David A. Lee
//
//Portions created by (your name) are Copyright (C) (your legal entity). All Rights Reserved.
//
//Contributor(s): none.
//
