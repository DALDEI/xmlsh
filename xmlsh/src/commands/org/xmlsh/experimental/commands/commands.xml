<?xml version="1.0"?>
<commands>

<command name="xobsufcate" url="http://www.xmlsh.org/CommandXobsufcate"/>
<command name="xsysinfo" url="http://www.xmlsh.org/CommandXsysinfo"/>
</commands>