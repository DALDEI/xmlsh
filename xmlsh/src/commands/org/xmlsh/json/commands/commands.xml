<commands>
 
   <function name="boolean" url="http://www.xmlsh.org/FunctionJsonBoolean">
		    <synopsis>construct a JSON boolean</synopsis>
	  </function>
	  
	     <function name="null" url="http://www.xmlsh.org/FunctionJsonNull">
		    <synopsis>construct a JSON null</synopsis>
	  </function>
	     <function name="array" url="http://www.xmlsh.org/FunctionJsonarray">
		    <synopsis>construct a JSON array</synopsis>
	  </function>
	     <function name="json" url="http://www.xmlsh.org/FunctionJsonJson">
		    <synopsis>construct a JSON object from a string</synopsis>
	  </function>
	     <function name="object" url="http://www.xmlsh.org/FunctionJsonObject">
		    <synopsis>construct a JSON object</synopsis>
	  </function>
	     <function name="boolean" url="http://www.xmlsh.org/FunctionJsonPath">
		    <synopsis>Evaluate a JsonPath on a JSON object</synopsis>
	  </function>
	  	     <function name="string" url="http://www.xmlsh.org/FunctionJsonString">
		    <synopsis>construct a JSON string</synopsis>
	  </function>
	  	     <function name="value" url="http://www.xmlsh.org/FunctionJsonValue">
		    <synopsis>construct a JSON value</synopsis>
	  </function>
	  	  </function>
	  	     <function name="number" url="http://www.xmlsh.org/FunctionJsonNumber">
		    <synopsis>construct a JSON number</synopsis>
	  </function>
</commands>
