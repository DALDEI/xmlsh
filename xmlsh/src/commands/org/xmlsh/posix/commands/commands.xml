<?xml version="1.0"?>
<commands>

<command name="base64" url="http://www.xmlsh.org/CommandBase64">
	<snopsis>Encode/Decode files to/from base64</snopsis>
</command>

<command name="cat" url="http://www.xmlsh.org/CommandCat">
	<snopsis>Copy (or concatenate) files</snopsis>
</command>
<command name="chmod" url="http://www.xmlsh.org/CommandChmod">
	<synopsis>Change file permissions</synopsis>
</command>
<command name="cp" url="http://www.xmlsh.org/CommandCp">
	<synopsis>Copy one or more files</synopsis>
</command>
<command name="ls" url="http://www.xmlsh.org/CommandLs">
	<synopsis>List files in a directory</synopsis>
</command>
<command name="mkdir" url="http://www.xmlsh.org/CommandMkdir">
	<synopsis>Create directory</synopsis>
</command>
<command name="more" url="http://www.xmlsh.org/CommandMore">
	<synopsis>Simple pagenator</synopsis>
</command>
<command name="mv" url="http://www.xmlsh.org/CommandMv">
	<synopsis>Move/Rename files</synopsis>
</command>
<command name="rm" url="http://www.xmlsh.org/CommandRm">
	<synopsis>Remove/Delete files</synopsis>
</command>
<command name="rmdir" url="http://www.xmlsh.org/CommandRmdir">
	<synopsis>Remove/Delete directories</synopsis>
</command>
<command name="sleep" url="http://www.xmlsh.org/CommandSleep">
	<synopsis>Delay execution</synopsis>
</command>
<command name="touch" url="http://www.xmlsh.org/CommandTouch" >
	<synopsis>Create (and update) timestamp of a file</synopsis>
</command>
<command name="tee" url="http://www.xmlsh.org/CommandTee" >
	<synopsis>Read from STDIN and write to STDOUT and files</synopsis>
</command>
<command name="mktemp" url="http://www.xmlsh.org/CommandMktemp" >
	<synopsis>Creates a temporary file</synopsis>
</command>

</commands>
	