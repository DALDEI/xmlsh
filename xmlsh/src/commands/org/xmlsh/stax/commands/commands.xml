
<commands>

	<command name="closeReader" url="http://www.xmlsh.org/CommandStAXcloseReader">
		<synopsis>Close a StAX reader</synopsis>
		<usage>
			<para>closeReader $reader</para>
		</usage>
	</command>

	<function name="getAttribute" url="http://www.xmlsh.org/FuncitonStAXgetAttribute">
		<synopsis>Get an attribute value from a StAX START_ELEMENT event</synopsis>
		<usage>
			<para>getAttribute( $event qname  )</para>
		</usage>
	</function>

	<function name="getData" url="http://www.xmlsh.org/FunctionStAXgetData">
		<synopsis>Get a StAX event’s data</synopsis>
		<usage>
			<para>getData($event)</para>
		</usage>
	</function>

	<function name="getEventType" url="http://www.xmlsh.org/FunctionStAXgetEventType">
		<synopsis>Get a StAX event’s type</synopsis>
		<usage>
			<para>getEventType($event)</para>
		</usage>
	</function>
	<function name="getName" url="http://www.xmlsh.org/FunctionStAXgetName">
		<synopsis>Get StAX event’s name (as QName)</synopsis>
		<usage>
			<para>getName($event)</para>
		</usage>
	</function>
	<function name="getNamespace" url="http://www.xmlsh.org/FunctionStAXgetNamespace">
		<synopsis>Get a StAX event’s namespace URI (as string)</synopsis>
		<usage>
			<para>getNamespace($event)</para>
		</usage>
	</function>

	<function name="hasNext" url="http://www.xmlsh.org/FunctionStAXhasNext">
		<synopsis>Advance the StAX stream to the next matching event</synopsis>
		<usage>
			<para>hasNext($reader)</para>
			<para>hasNext($reader event_type)</para>
			<para>hasNext($reader event_type qname)</para>
		</usage>
	</function>

	<function name="nextEvent" url="http://www.xmlsh.org/FunctionStAXnextEvent">
		<synopsis>Return the next event</synopsis>
		<usage>
			<para>nextEvent($reader)</para>
			<para>nextEvent($reader event_type)</para>
			<para>nextEvent($reader event_type qname)</para>
		</usage>
	</function>
	<function name="newEventReader" url="http://www.xmlsh.org/FunctionStAXnewEventReader">
		<synopsis>Create a new XMLEventReader</synopsis>
		<usage>
			<para>newEventReader()</para>
			<para>newEventReader(uri)</para>
		</usage>
	</function>
</commands>