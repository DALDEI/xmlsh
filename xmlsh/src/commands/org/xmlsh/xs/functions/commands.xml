<commands>
 
   <function name="string" url="http://www.xmlsh.org/FunctionXSString">
		    <synopsis>Convert to string</synopsis>
		    <usage>
			      <para>string( args )</para>
		    </usage>
	  </function>
   
   <function name="integer" url="http://www.xmlsh.org/FunctionXSInteger">
		    <synopsis>Convert to integer</synopsis>
		    <usage>
			      <para>integer( arg )</para>
		    </usage>
	  </function>
	    <function name="decimal" url="http://www.xmlsh.org/FunctionXSInteger">
		    <synopsis>Convert to integer</synopsis>
		    <usage>
			      <para>decimal( arg )</para>
		    </usage>
	  </function>
	     <function name="attribute" url="http://www.xmlsh.org/FunctionXSAttribute">
		    <synopsis>constructs an attribute</synopsis>
		    <usage>
			      <para>attribute( name value )</para>
		    </usage>
	  </function>
	  	     <function name="elemement" url="http://www.xmlsh.org/FunctionXSElement">
		    <synopsis>constructs an element</synopsis>
		    <usage>
			      <para>element( name value [value ...] )</para>
		    </usage>
	  </function>
	  	     <function name="document" url="http://www.xmlsh.org/FunctionXSDocument">
		    <synopsis>constructs a document</synopsis>
		    <usage>
			      <para>document( root )</para>
		    </usage>
	  </function>
	  
</commands>
