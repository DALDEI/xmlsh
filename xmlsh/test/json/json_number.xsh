# test json:boolean
import commands json=json

echo json:number( 1 )
echo json:number( 1.5 )
echo json:number( -1 )
echo json:object( a json:number( 2 ) c json:number( 3.5 ) )



