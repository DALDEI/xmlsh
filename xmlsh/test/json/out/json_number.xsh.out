1
1.5
-1
{"a":2,"c":3.5}
